commit 0f9eb85794c7b812801a7f43e245aee986cb38a5
Author: Bron Gondwana <brong@opera.com>
Date:   Wed Nov 21 11:41:04 2012 +0100

    Bug #3754 - parse GUID for binary appends as well

diff --git a/imap/message.c b/imap/message.c
index e97be01..4daddf9 100644
--- a/imap/message.c
+++ b/imap/message.c
@@ -371,6 +371,8 @@ int message_parse_binary_file(FILE *infile, struct body **body)
     message_parse_body(&msg, MAILBOX_FORMAT_NORMAL, *body,
 		       DEFAULT_CONTENT_TYPE, (struct boundary *)0);
 
+    message_guid_generate(&(*body)->guid, msg.base, msg.len);
+
     lseek(fd, 0L, SEEK_SET);
     n = retry_write(fd, msg.base, msg.len);
 
